central/group/datastore/datastore_impl.go

```diff
@@ -27,6 +27,57 @@ type dataStoreImpl struct {
 	lock sync.RWMutex
 }
 
+func (ds *dataStoreImpl) Upsert(ctx context.Context, group *storage.Group) error {
+	if ok, err := accessSAC.WriteAllowed(ctx); err != nil {
+		return err
+	} else if !ok {
+		return sac.ErrResourceAccessDenied
+	}
+
+	// Lock to simulate being behind a transaction
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	if err := ds.validateAndPrepGroupForUpsertNoLock(ctx, group); err != nil {
+		return err
+	}
+
+	return ds.storage.Upsert(ctx, group)
+}
+
+// Validate if the group is allowed to be upserted and prep the group before it is upserted in db.
+// NOTE: This function assumes that the call to this function is already behind a lock.
+func (ds *dataStoreImpl) validateAndPrepGroupForUpsertNoLock(ctx context.Context, newGroup *storage.Group) error {
+	if err := ValidateGroup(newGroup, true); err != nil {
```

Comment:
Shouldn't we only require a ID _iff_ the group didn't exist beforehand? Or do we want to proclaim that for `Upsert`, the ID must be set at all times?

Reply:
I think that depends on whether we want to treat `Upsert` as the function that can be used outside the declarative context or not.

If it were to be used outside the declarative context, then we shouldn't require ID. If we only use this function within declarative context, we know for sure group will have deterministic ID set

Reply:
I suppose there's nothing stopping anyone using the datastore's generic interface to use the `Upsert` function currently.

The cleanest solution would be IMO to make it generic, and also usable within the non-declarative context. I don't see why a generic interface has a specific function only usable in a specific context - if we make it part of the generic interface, it should serve a generic purpose.